risc0/cargo-risczero/src/commands/build_guest.rs

```diff
@@ -0,0 +1,201 @@
+            .run(r"find target/riscv32im-risc0-zkvm-elf/release -maxdepth 1 -type f -exec test -x {} \; -exec cp {} /tmp/ \;");
+
+        let binary: DockerFile<'_> = DockerFile::new()
+            .comment("binary stage")
+            .from_alias("binary", "scratch")
+            .env(pkg_env)
+            .copy_from("build", "/tmp", "/$PKG_NAME");
+
+        let file = DockerFile::new().dockerfile(build).dockerfile(binary);
+        let mut dockerfile = File::create("Dockerfile")?;
+        dockerfile.write_all(file.to_string().as_bytes())?;
+        let mut dockerignore = File::create(".dockerignore")?;
+        dockerignore.write_all(DOCKER_IGNORE.as_bytes())?;
+
+        Ok(())
+    }
+
+    /// Build the dockerfile and ouputs the ELF.
+    ///
+    /// Overwrites if an ELF with the same name already exists.
+    fn build(&self) -> Result<()> {
+        Command::new("docker")
+            .args(["build", "--output=elfs/", "--target=binary", "."])
+            .run_verbose()?;
+
+        Ok(())
+    }
+
+    /// Remove Dockerfile and .dockerignore files.
+    fn clean(&self) -> Result<()> {
+        Command::new("rm")
+            .args(["Dockerfile", ".dockerignore"])
+            .run_verbose()
+    }
+
+    fn check_cargo_lock(&self) -> Result<()> {
+        let lock_file = PathBuf::from(&self.manifest_path)
+            .parent()
+            .context("invalid manifest path")?
+            .join("Cargo.lock");
+        fs::metadata(lock_file.clone()).context(format!(
+            "Cargo.lock not found in path {}",
+            lock_file.display()
+        ))?;
+        Ok(())
+    }
+
+    /// Compute the image ID for a given ELF.
+    fn image_id(&self, elf_path: &PathBuf) -> Result<Digest> {
+        let elf = fs::read(elf_path)?;
+        let program = Program::load_elf(&elf, MEM_SIZE as u32).context("unable to load elf")?;
+        let image = MemoryImage::new(&program, PAGE_SIZE as u32)
+            .context("unable to create memory image")?;
+        Ok(image.compute_id())
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use std::{env, path::Path};
+
+    use super::BuildGuest;
+
+    struct Tester {
+        manifest_path: String,
+        elf_path: String,
+        expected_image_id: String,
+    }
+
+    impl Tester {
+        fn run(&self) {
+            env::set_current_dir("../../").unwrap();
+            let builder = BuildGuest {
+                manifest_path: self.manifest_path.clone(),
+            };
+            builder.run().unwrap();
+            assert_eq!(
+                builder
+                    .image_id(&Path::new(&self.elf_path).to_path_buf())
+                    .unwrap()
+                    .to_string(),
+                self.expected_image_id
+            );
+        }
+    }
+    #[test]
+    #[ignore] // requires Docker to be installed
+    fn test_reproducible_multiply_method() {
+        let multiply_test = Tester {
+            manifest_path: "examples/factors/methods/guest/Cargo.toml".to_string(),
```

Comment:
Something a bit more persistent would be `risc0/zkvm/methods/guest/Cargo.toml`

Reply:
We are not committing the Cargo.lock for that method though